### PR TITLE
E2E Test: New module to test the integration with Yara

### LINKED ISSUE #3085: E2E design and implementation: Yara
## Description

The test research was accomplished in [this issue](https://github.com/wazuh/wazuh-qa/issues/2860). So, in this issue, we will design and develop the E2E test to cover those use cases.

## Tasks

- [x] Design the test
- [x] Implement the test locally
- [ ] Refactor the test for it to work in the test environment

## Conclusion

WIP.

### PR DESCRIPTION
**Pull request by mauromalara**
|Related issue|
|-------------|
| Closes #3085 |

## Description

The added module tests the integration with Yara. The only precondition to run is to create an inventory as follows:
```
all:
  hosts:
    wazuh-manager:
      ansible_connection: ssh
      ansible_user: <VM USER>
      ansible_ssh_private_key_file: <PRIVATE KEY PATH>
      ansible_python_interpreter: <PYTHON INTERPRETER PATH>
      dashboard_user: <USER>
      dashboard_password: <PASS>
```

The test structure is the following:

```
tests/end_to_end/test_yara_integration/
├── data
│   ├── configuration
│   │   ├── malware_downloader.sh
│   │   └── yara.sh
│   ├── playbooks
│   │   ├── configuration.yaml
│   │   ├── generate_events.yaml
│   │   └── teardown.yaml
│   └── test_cases
│       └── cases_yara_integration.yaml
└── test_yara_integration.py
```

### Added

+ `configuration`: this folder contains all necessary scripts to generate the events that must trigger alerts in the manager.
+ `playbooks`: this folder contains all necessary Ansible playbooks to (1) prepare the environment, (2) generate the events and get alerts, and (3) clean the environment.
+ `test_cases`: this folder contains a YAML file with the test cases that will be executed in this test (in this stage we will only create 1 test case).
+ `test_yara_integration.py`: this is the module that test the integration.

### Updated

+ `end_to_end/conftest.py`: now it is possible to send module extra vars to the test case playbook.

---

## Testing performed [Code analysis](https://github.com/wazuh/wazuh-qa/runs/7567459794?check_suite_focus=true)

| Tester             | Test path | Jenkins | Local  | OS | Commit | Notes                |
|--------------------|-----------|---------|--------|-----|--------|----------------------|
| @mauromalara (Developer) | end_to_end/test_yara_integration | n/a | [🟢](https://github.com/wazuh/wazuh-qa/files/9214053/r1-CentOS-manager-3085.zip) [🟢](https://github.com/wazuh/wazuh-qa/files/9214054/r2-CentOS-manager-3085.zip) [🟢](https://github.com/wazuh/wazuh-qa/files/9214055/r3-CentOS-manager-3085.zip) | CentOS | fa3765a | Nothing to highlight |
| @juliamagan  (Reviewer)   |   test_yara_integration/        | :no_entry_sign: :no_entry_sign: :no_entry_sign:|[🟢 ](https://github.com/wazuh/wazuh-qa/files/9233355/R1-3115-test-yara-integration.zip)[🟢 ](https://github.com/wazuh/wazuh-qa/files/9233357/R2-3115-test-yara-integration.zip)[🟢 ](https://github.com/wazuh/wazuh-qa/files/9233359/R3-3115-test-yara-integration.zip) |    CentOS    |    fa3765a   | Nothing to highlight |

### CODE REVIEW

File: tests/end_to_end/test_yara_integration/data/playbooks/teardown.yaml [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+      file:
+        path: /tmp/yara_script.sh
+        state: absent
+
+    - name: Remove jq (JSON Processor) on CentOS
+      package:
+        name:
+          - jq
+        state: absent
+      when: ansible_facts['distribution'] == "CentOS"
```
Comment (juliamagan) [inline tests/end_to_end/test_yara_integration/data/playbooks/teardown.yaml:47]: This is not needed.

File: tests/end_to_end/test_yara_integration/test_yara_integration.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+          "term": {
+            "rule.level": f"{rule_level}"
+          }
+        },
+        {
+          "term": {
+            "data.yara_rule": f"{data_yara_rule}"
+          }
+        }
+    ])
```
Comment (juliamagan) [inline tests/end_to_end/test_yara_integration/test_yara_integration.py:63]: Get the alert with the timestamp. You can check other tests in the branch.

File: tests/end_to_end/conftest.py [hunk 3]
```diff
@@ -105,9 +105,16 @@ def generate_events(request, metadata):
...
+
         # Check if the test case has extra variables to pass to the playbook and add them to the parameters in that case
+        # Also, add the module extra vars if it is configured.
+        module_extra_vars = getattr(request.module, 'configuration_extra_vars', None)
         if 'extra_vars' in metadata:
             parameters.update({'extravars': metadata['extra_vars']})
+            if module_extra_vars is not None:
+                parameters['extravars'].update(module_extra_vars)
+        elif module_extra_vars is not None:
+            parameters.update({'extravars': module_extra_vars})
```
Comment (juliamagan) [inline tests/end_to_end/conftest.py:117]: Extra vars for generating events should be in the test case metadata.

**Review by juliamagan — CHANGES_REQUESTED** [review]
GJ, but some changes are required

File: tests/end_to_end/test_yara_integration/data/playbooks/teardown.yaml (hunk 1 above)
Comment (mauromalara) [reply tests/end_to_end/test_yara_integration/data/playbooks/teardown.yaml:47]: Done in 2a5ea15

File: tests/end_to_end/conftest.py (hunk 3 above)
Comment (mauromalara) [reply tests/end_to_end/conftest.py]: Done in [df2a57b](https://github.com/wazuh/wazuh-qa/pull/3115/commits/df2a57b36e82d86c756ab4da90e286cc955d3799)

File: tests/end_to_end/test_yara_integration/test_yara_integration.py (hunk 2 above)
Comment (mauromalara) [reply tests/end_to_end/test_yara_integration/test_yara_integration.py]: Done in [df2a57b](https://github.com/wazuh/wazuh-qa/pull/3115/commits/df2a57b36e82d86c756ab4da90e286cc955d3799)

File: tests/end_to_end/test_yara_integration/data/test_cases/cases_yara_integration.yaml [hunk 4]
```diff
@@ -0,0 +1,9 @@
+- name: Run Yara scan to detect malware
+  description: Test the detection of a trojaned file
+  configuration_parameters: null
+  metadata:
+    rule.id: 108001
+    rule.level: 12
+    rule.description: 'File \\"/tmp/yara/malware/mirai\\" is a positive match. Yara rule: Mirai_Botnet_Malware_RID2EF6'
+    extra:
```
[the remark this answers is not in the document]
Comment (mauromalara) [reply tests/end_to_end/test_yara_integration/data/test_cases/cases_yara_integration.yaml:9]: Done in [df2a57b](https://github.com/wazuh/wazuh-qa/pull/3115/commits/df2a57b36e82d86c756ab4da90e286cc955d3799)

File: tests/end_to_end/test_yara_integration/data/playbooks/teardown.yaml (hunk 1 above)
Comment (mauromalara) [reply tests/end_to_end/test_yara_integration/data/playbooks/teardown.yaml]: Done in 2a5ea15

**Review by juliamagan — APPROVED** [review]

**Review by jmv74211 — CHANGES_REQUESTED** [review]

File: tests/end_to_end/test_yara_integration/test_yara_integration.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+    expected_alert_json = fr".+timestamp\":\"({timestamp_regex})\",.+level\":{rule_level}.+description\"" \
+                          fr":\"{rule_description}.+id.+{rule_id}"
+
+    expected_indexed_alert = fr".+yara_rule\": \"{data_yara_rule}.+level.+{rule_level}.+id.+{rule_id}.+" \
+                             fr"timestamp\": \"({timestamp_regex})\""
+
+    # Check that alert has been raised and save timestamp
+    raised_alert = evm.check_event(callback=expected_alert_json, file_to_monitor=alerts_json,
+                                   error_message='The alert has not occurred').result()
```
Comment (jmv74211) [inline tests/end_to_end/test_yara_integration/test_yara_integration.py:53]:
```suggestion
                                   timeout=1, error_message='The alert has not occurred').result()
```

File: tests/end_to_end/test_yara_integration/data/configuration/yara.sh [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+read INPUT_JSON
+YARA_PATH=$(echo $INPUT_JSON | jq -r .parameters.extra_args[1])
+YARA_RULES=$(echo $INPUT_JSON | jq -r .parameters.extra_args[3])
+FILENAME=$(echo $INPUT_JSON | jq -r .parameters.alert.syscheck.path)
+
+# Set LOG_FILE path
+LOG_FILE="logs/active-responses.log"
+
+size=0
+actual_size=$(stat -c %s ${FILENAME})
```
[the remark this answers is not in the document]
Comment (mauromalara) [reply tests/end_to_end/test_yara_integration/data/configuration/yara.sh:23]: This script was extracted from the PoC guide, so an explanation is not required in this case.

File: tests/end_to_end/test_yara_integration/data/configuration/yara.sh [hunk 7]
```diff
@@ -0,0 +1,51 @@
+#!/bin/bash
+# Wazuh - Yara active response
+# Copyright (C) 2015-2022, Wazuh Inc.
+#
+# This program is free software; you can redistribute it
+# and/or modify it under the terms of the GNU General Public
+# License (version 2) as published by the FSF - Free Software
+# Foundation.
+
```
[the remark this answers is not in the document]
Comment (mauromalara) [reply tests/end_to_end/test_yara_integration/data/configuration/yara.sh:9]: This script was extracted from the PoC guide, so an explanation is not required in this case.

File: tests/end_to_end/test_yara_integration/test_yara_integration.py (hunk 5 above)
Comment (mauromalara) [reply tests/end_to_end/test_yara_integration/test_yara_integration.py]: Done in 97856db

File: tests/end_to_end/test_yara_integration/data/playbooks/configuration.yaml [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+    - name: Check if jq is installed
+      shell: rpm -qa jq
+      register: check_jq
+
+    - name: Install jq (JSON Processor) on CentOS
+      package:
+        name:
+          - jq
+        state: present
+      when: ( ansible_facts['distribution'] == "CentOS" and "jq" not in check_jq.stdout)
```
[the remark this answers is not in the document]
Comment (mauromalara) [reply tests/end_to_end/test_yara_integration/data/playbooks/configuration.yaml]: Done in 97856db

File: tests/end_to_end/test_yara_integration/data/playbooks/configuration.yaml [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+    - name: Check if epel-release is installed
+      shell: rpm -qa epel-release
+      register: check_epel_release
+
+    - name: Install epel-release
+      package:
+        name:
+          - epel-release
+        state: present
+      when: ( ansible_facts['distribution'] == "CentOS" and "epel" not in check_epel_release.stdout)
```
[the remark this answers is not in the document]
Comment (mauromalara) [reply tests/end_to_end/test_yara_integration/data/playbooks/configuration.yaml]: Done in 97856db

File: tests/end_to_end/test_yara_integration/data/configuration/malware_downloader.sh [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+  curl -s -XGET "$1" -o "$2"
+
+}
+
+echo
+# Mirai
+echo "# Mirai: https://en.wikipedia.org/wiki/Mirai_(malware)"
+echo "Downloading malware sample..."
+fetch_sample "https://wazuh-demo.s3-us-west-1.amazonaws.com/mirai" "/tmp/yara/malware/mirai" && echo "Done!" || echo "Error while downloading."
+echo
```
[the remark this answers is not in the document]
Comment (mauromalara) [reply tests/end_to_end/test_yara_integration/data/configuration/malware_downloader.sh]: Done in 97856db

File: deps/wazuh_testing/wazuh_testing/end_to_end/__init__.py [hunk 11]
```diff
@@ -47,20 +50,12 @@ def delete_index_api(credentials, ip_address='wazuh-manager', index='wazuh-alert
...
-    assert response.status_code == HTTPStatus.OK, 'The index(es) have not been obtained successfully.' \
-                                                  f"Actual response {response.status_code}"
-
-    if response.text == '':
-        return None
-
     response = requests.delete(url=url+index, params={'pretty': 'true'}, verify=False, auth=authorization)
-    assert response.status_code == HTTPStatus.OK, 'The index(es) have not been deleted successfully.' \
-                                                  f"Actual response {response.status_code}"
+    assert response.status_code == HTTPStatus.OK, 'The index(es) have not been deleted successfully. ' \
```
[the remark this answers is not in the document]
Comment (mauromalara) [reply deps/wazuh_testing/wazuh_testing/end_to_end/__init__.py]: Done in 8065c8d

File: deps/wazuh_testing/wazuh_testing/end_to_end/__init__.py [hunk 12]
```diff
@@ -29,6 +29,9 @@ def get_alert_indexer_api(query, credentials, ip_address='wazuh-manager', index=
                             auth=requests.auth.HTTPBasicAuth(credentials['user'], credentials['password']))
     if '"hits" : [ ]' in response.text:
         raise Exception('Alert not indexed')
+    else:
+        assert response.status_code == HTTPStatus.OK, "The request wasn't successful. " \
```
[the remark this answers is not in the document]
Comment (mauromalara) [reply deps/wazuh_testing/wazuh_testing/end_to_end/__init__.py]: Done in 8065c8d

File: tests/end_to_end/test_yara_integration/test_yara_integration.py [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+
+    expected_alert_json = fr".+timestamp\":\"({timestamp_regex})\",.+level\":{rule_level}.+description\"" \
+                          fr":\"{rule_description}.+id.+{rule_id}"
+
+    expected_indexed_alert = fr".+yara_rule\": \"{data_yara_rule}.+level.+{rule_level}.+id.+{rule_id}.+" \
+                             fr"timestamp\": \"({timestamp_regex})\""
+
+    # Check that alert has been raised and save timestamp
+    raised_alert = evm.check_event(callback=expected_alert_json, file_to_monitor=alerts_json,
+                                   timeout=5, error_message='The alert has not occurred').result()
```
Comment (juliamagan) [inline tests/end_to_end/test_yara_integration/test_yara_integration.py:48]: Hardcoded timeouts are not allowed

**Review by juliamagan — CHANGES_REQUESTED** [review]

File: tests/end_to_end/test_yara_integration/test_yara_integration.py (hunk 13 above)
Comment (mauromalara) [reply tests/end_to_end/test_yara_integration/test_yara_integration.py]: Done in fa3765a

**Review by juliamagan — APPROVED** [review]

**Review by jmv74211 — APPROVED** [review]

**Pull request merged**